New zee package

tristan0x opened the pull request:
This pull-request provides a new zee package as well improvements and addition of all requirements packages.

Commit:
gmodel: new package starting at 2.1.0

Commit:
gmsh: add git master version

Commit:
omega-h: add versions and variants

* add version 9.13.14 and git master branch
* add trilinos, gmodel, throw, and examples variants
* disable XSDK so that Trilinos dependencies are found.

Commit:
zee: new package

pramodk commented on var/spack/repos/builtin/packages/gmsh/package.py:
```diff
@@ -44,6 +44,7 @@ class Gmsh(CMakePackage):
     version('2.15.0', '992a4b580454105f719f5bc05441d3d392ab0b4b80d4ea07b61ca3bdc974070a')
     version('2.12.0', '7fbd2ec8071e79725266e72744d21e902d4fe6fa9e7c52340ad5f4be5c159d09')
     version('2.11.0', 'f15b6e7ac9ca649c9a74440e1259d0db')
+    version('master', branch='master', git='https://gitlab.onelab.info/gmsh/gmsh.git')
```
I would suggest to use `develop` version name, its common convention in spack packages. Also, there is special handling [rules](https://spack.readthedocs.io/en/latest/packaging_guide.html#version-comparison) for it.

pramodk commented on var/spack/repos/builtin/packages/omega-h/package.py:
```diff
@@ -34,18 +34,31 @@ class OmegaH(CMakePackage):
+    git      = "https://github.com/ibaned/omega_h.git"
 
-    version('9.13.4', '035f9986ec07ad97ae0aa1e171872307')
+    version('master', branch='master')
```
same here : master->develop

pramodk commented on var/spack/repos/builtin/packages/omega-h/package.py:
```diff
@@ -34,18 +34,31 @@ class OmegaH(CMakePackage):
+    variant('examples', default=False, description='Compile examples')
 
+    depends_on('gmodel', when='+gmodel')
+    depends_on('gmsh', when='+examples', type='build')
```
just to be sure: gmsh is not linked?

tristan0x replied:
Indeed gmsh is not linked. It is used at build time to create .msh from .geo

ohm314 replied:
it's actually not even really a requirement for build - only for running the tests, which is in the CMake files. Maybe there should be a specific target for this in the zee cmake.

pramodk commented on var/spack/repos/builtin/packages/omega-h/package.py:
```diff
@@ -34,18 +34,31 @@ class OmegaH(CMakePackage):
     def cmake_args(self):
-        args = ['-DUSE_XSDK_DEFAULTS:BOOL=ON']
+        args = ['-DUSE_XSDK_DEFAULTS:BOOL=OFF']
         if '+shared' in self.spec:
```
I assume this default OFF will be in BBP specific repo?

pramodk commented on var/spack/repos/builtin/packages/zee/package.py:
```diff
@@ -0,0 +1,61 @@
+    homepage = "https://github.com/BlueBrain/zee"
+    url      = "git@github.com:BlueBrain/zee.git"
+
+    version('master', git=url, submodules=True)
```
master->develop

pramodk commented on var/spack/repos/builtin/packages/zee/package.py:
```diff
@@ -0,0 +1,61 @@
+        return flags
+
+    def cmake_args(self):
+        spec   = self.spec
```
spec not used

Commit:
petsc: add patch to fix MPI setup with C++11 and 64bits integers

PETSc implementation wrongly claims that `MPI_INT64_T` should not be used when not a macro.

tristan0x commented:
Thank you for the feedback @pramodk 
I pushed new changes to take in remarks into accounts.

I also pushed a patch for PETSc to properly use it with Omega_h arrays.
The patch works at least for HPE-MPI and Intel compiler. I will ensure it works in more situations later on.

pramodk reviewed (approved):
LGTM. 

> I also pushed a patch for PETSc to properly use it with Omega_h arrays.
> The patch works at least for HPE-MPI and Intel compiler. I will ensure it works in more situations later on.

I added comment BlueBrain/zee/pull/15. If this is reasonable for you guys then go ahead with merge.

Commit:
bb5 packages: add concretization preferences

ohm314 commented on var/spack/repos/builtin/packages/petsc/mpi_int64_t_c++11.diff:
```diff
@@ -0,0 +1,13 @@
+ .seealso: PetscReal, PetscScalar, PetscComplex, PetscInt, MPIU_REAL, MPIU_SCALAR, MPIU_COMPLEX
+ M*/
+ 
+-#if defined(PETSC_HAVE_STDINT_H) && defined(PETSC_HAVE_INTTYPES_H) && defined(PETSC_HAVE_MPI_INT64_T) /* MPI_INT64_T is not guaranteed to be a macro */
```
@pramodk made a good point that we should be careful with this as it might affect other dependencies of/in petsc (e.g. python bindings). I am a bit hesitant whether we should push this diff before it was not approved by PETSc devs

tristan0x replied:
This patch is only based on my reading of the headers of mpich 3.2 and HPE-MPI. I want to test against other implementations before contacting PETSc.
 
I think it is more urgent to start evaluating zee performance right now than making sure that the patch works on all super-computers of the world.

ohm314 commented on var/spack/repos/builtin/packages/zee/package.py:
```diff
@@ -0,0 +1,60 @@
+    depends_on('gmsh -mpi %gcc')
+    depends_on('mpi')
+    depends_on('omega-h')
+    depends_on('petsc +int64')
```
if above changes to petsc are to be taken into accout, don't we need the variant for `patchmpi64 ` here?

tristan0x replied:
For now I want to apply the patch on demand, so that I can test easily other MPI implementations without.

tristan0x commented:
> it's actually not even really a requirement for build - only for running the tests, which is in the CMake files. Maybe there should be a specific target for this in the zee cmake.

From what I understand, spack `build` phase includes configure, build, and tests of the package. So gmsh is required and only required during this spack `build` phase.

pramodk commented:
This is ok to merge. Only remaining question I had was `+debug` variant. Why we can't use `build_type` variant already present for all CMake package.

tristan0x commented:
> I had was +debug variant. Why we can't use build_type variant already present for all CMake package.

I think the `+debug` was in the spack template. I thought it was kind of standard. I remove it right away!
